### PR TITLE
core: check if sender is EOA

### PR DESCRIPTION
**Pull request by holiman**
This is a rebased version of #23002. Some commits are dropped, and I didn't want to squash-delete those until I'm sure they're not needed, hence the new PR.

### CODE REVIEW

File: core/state_transition.go [hunk 1]
```diff
@@ -220,6 +223,11 @@ func (st *StateTransition) preCheck() error {
 				st.msg.From().Hex(), msgNonce, stNonce)
 		}
 	}
+	// Make sure the sender is an EOA
+	if codeHash := st.state.GetCodeHash(st.msg.From()); codeHash != emptyCodeHash && codeHash != (common.Hash{}) {
+		return fmt.Errorf("%w: address %v, codehash: %s", ErrSenderNoEOA,
```
Comment (rjl493456442) [inline core/state_transition.go:228]: Please use "%w: address %v, codehash: %x"

Comment (holiman) [reply core/state_transition.go:228]: If we use `%x` on `st.mst.From()`, we don't get the checksummed address. IMO it's better the way it is. As for the codehash, `%x` makes no difference, aside from the `0x` prefix being dropped if we do `%x`.

Comment (rjl493456442) [reply core/state_transition.go:228]: Yes right. I was wrong here.

**Review by rjl493456442 — APPROVED** [review]
LGTM!

File: core/state_transition.go [hunk 2]
```diff
@@ -220,6 +223,11 @@ func (st *StateTransition) preCheck() error {
 				st.msg.From().Hex(), msgNonce, stNonce)
 		}
 	}
+	// Make sure the sender is an EOA
+	if codeHash := st.state.GetCodeHash(st.msg.From()); codeHash != emptyCodeHash && codeHash != (common.Hash{}) {
```
Comment (karalabe) [inline core/state_transition.go:227]: Why did you check against the zero hash? That should never happen.

**Comment by holiman** [comment]
See https://github.com/ethereum/go-ethereum/pull/23002#issuecomment-872282874

**Pull request merged**